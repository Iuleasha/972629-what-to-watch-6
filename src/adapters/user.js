export const userAdapter = ({
  id = ``,
  email = ``,
  name = ``,
  avatar_url: avatarUrl = ``,
}) => {
  return {
    id: id.toString(),
    email,
    name,
    avatarUrl,
  };
};
