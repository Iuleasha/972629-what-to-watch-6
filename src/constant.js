export const HeaderMode = {
  MOVIE_CARD: `movie-card__head`,
  USER_PAGE: `user-page__head`,
};

export const PosterSize = {
  WIDTH: 218,
  HEIGHT: 327,
};

export const MovieCardButtonSize = {
  WIDTH: 19,
  HEIGHT: 20,
};

export const MOVIE_CARD_NAV_ITEMS = [`Overview`, `Details`, `Reviews`];

export const DEFAULT_GENRE = `All genres`;

export const AppRoute = {
  LOGIN: `/login`,
  LOSE: `/lose`,
  MY_LIST: `/mylist`,
  ROOT: `/`,
  FILMS: `/films`,
  FILM: `/films/:id`,
  PLAYER: `/player/:id`,
  REVIEW: `/films/:id/review`,
};

export const AuthorizationStatus = {
  AUTH: `AUTH`,
  NO_AUTH: `NO_AUTH`,
};

export const APIRoute = {
  FILMS: `/films`,
  FILM: `/film`,
  PROMO_FILM: `/films/promo`,
  LOGIN: `/login`,
  LOGOUT: `/logout`,
  COMMENTS: `/comments`,
  FAVORITE: `/favorite`
};
