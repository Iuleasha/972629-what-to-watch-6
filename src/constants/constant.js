export const HeaderMode = {
  MOVIE_CARD: `movie-card__head`,
  USER_PAGE: `user-page__head`,
};

export const PosterSize = {
  WIDTH: 218,
  HEIGHT: 327,
};

export const MovieCardButtonSize = {
  WIDTH: 19,
  HEIGHT: 20,
};

export const MOVIE_CARD_NAV_ITEMS = [`Overview`, `Details`, `Reviews`];

export const DEFAULT_GENRE = `All genres`;

export const AppRoute = {
  LOGIN: `/login`,
  LOSE: `/lose`,
  MY_LIST: `/mylist`,
  ROOT: `/`,
  FILMS: `/films`,
  FILM: `/films/:id`,
  PLAYER: `/player/:id`,
  REVIEW: `/films/:id/review`,
};

export const AuthorizationStatus = {
  AUTH: `AUTH`,
  NO_AUTH: `NO_AUTH`,
};

export const APIRoute = {
  FILMS: `/films`,
  FILM: `/film`,
  PROMO_FILM: `/films/promo`,
  LOGIN: `/login`,
  LOGOUT: `/logout`,
  COMMENTS: `/comments`,
  FAVORITE: `/favorite`
};

export const ButtonIcon = {
  PLAY: `#play-s`,
  PAUSE: `#pause`,
  FULL_SCREEN: `#full-screen`,
  INLIST: `#in-list`,
  ADD: `#add`,
};

export const NameSpace = {
  DATA: `DATA`,
  COMMENTS: `COMMENTS`,
  USER: `USER`,
};

export const ActionType = {
  LOAD_SUCCESS: `load/loadSuccess`,
  LOAD_PROMO_FILM: `load/loadPromoFilm`,
  SET_GENRE: `filter/genres`,
  LOAD_FAVORITE: `load/favorite`,
  SWITCH_FAVORITE_STATUS: `load/switchFavoriteStatus`,
  REQUIRED_AUTHORIZATION: `user/requiredAuthorization`,
  REDIRECT_TO_ROUTE: `site/redirectToRoute`,
  SET_USER: `user/setUser`,
  ADD_REVIEW: `film/addReview`,
};

export const FILMS_PAGE_SIZE = 8;

export const AVATAR_SIZE = 63;

export const CommentLength = {
  MIN: 50,
  MAX: 400
};

export const STARS_LENGTH = 10;

export const RATING_DEFAULT_VALUE = 3;

export const RATING_STARS = [...Array(STARS_LENGTH).keys()].map((i) => i + 1);

export const Rating = {
  BAD: `Bad`,
  NORMAL: `Normal`,
  GOOD: `Good`,
  VERY_GOOD: `Very good`,
  AWESOME: `Awesome`,
};
