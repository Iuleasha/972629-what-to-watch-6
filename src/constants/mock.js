export const FILM_MOCK = {
  'id': 1,
  'name': `The Grand Budapest Hotel`,
  'poster_image': `img/the-grand-budapest-hotel-poster.jpg`,
  'preview_image': `img/the-grand-budapest-hotel.jpg`,
  'background_image': `img/the-grand-budapest-hotel-bg.jpg`,
  'background_color': `#ffffff`,
  'video_link': `https://some-link`,
  'preview_video_link': `https://some-link`,
  'description': `In the 1930s, the Grand Budapest Hotel is a popular European ski resort, presided over by concierge Gustave H. (Ralph Fiennes). Zero, a junior lobby boy, becomes Gustave's friend and protege.`,
  'rating': 8.9,
  'scores_count': 240,
  'director': `Wes Andreson`,
  'starring': [`Bill Murray`, `Edward Norton`, `Jude Law`, `Willem Dafoe`, `Saoirse Ronan`],
  'run_time': 99,
  'genre': `Comedy`,
  'released': 2014,
  'is_favorite': false,
};

export const COMMENT_MOCK = {
  'id': 1,
  'user': {
    'id': 4,
    'name': `Kate Muir`,
  },
  'rating': 8.9,
  'comment': `Discerning travellers and Wes Anderson fans will luxuriate in the glorious Mittel-European kitsch of one of the director's funniest and most exquisitely designed movies in years.`,
  'date': `2019-05-08T14:13:56.569Z`,
};

export const AUTH_INFO_MOCK = {
  'id': 1,
  'email': `Oliver.conner@gmail.com`,
  'name': `Oliver.conner`,
  'avatar_url': `img/1.png`,
};
