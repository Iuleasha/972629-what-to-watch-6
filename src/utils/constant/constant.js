export const HeaderMode = {
  MOVIE_CARD: `movie-card__head`,
  USER_PAGE: `user-page__head`,
};

export const PosterSize = {
  WIDTH: 218,
  HEIGHT: 327,
};

export const MovieCardButtonSize = {
  WIDTH: 19,
  HEIGHT: 20,
};

export const DEFAULT_GENRE = `All genres`;
