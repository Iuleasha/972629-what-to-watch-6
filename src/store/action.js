import {createAction} from '@reduxjs/toolkit';
import {ActionType} from '../constants/constant';

const getPayload = (payload) => ({payload});

export const setGenre = createAction(ActionType.SET_GENRE, getPayload);

export const loadFilms = createAction(ActionType.LOAD_SUCCESS, getPayload);

export const loadPromoFilm = createAction(ActionType.LOAD_PROMO_FILM, getPayload);

export const requireAuthorization = createAction(ActionType.REQUIRED_AUTHORIZATION, getPayload);

export const redirectToRoute = createAction(ActionType.REDIRECT_TO_ROUTE, getPayload);

export const setUser = createAction(ActionType.SET_USER, getPayload);

export const addReview = createAction(ActionType.ADD_REVIEW, getPayload);

export const loadFavorite = createAction(ActionType.LOAD_FAVORITE, getPayload);

export const switchFavoriteStatus = createAction(ActionType.SWITCH_FAVORITE_STATUS, getPayload);
