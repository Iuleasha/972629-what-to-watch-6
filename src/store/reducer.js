import {AuthorizationStatus, DEFAULT_GENRE} from '../constant';
import {createGenreArray} from '../filter/filter';
import {ActionType} from './action';

const initialState = {
  genre: DEFAULT_GENRE,
  genres: [],
  films: [],
  authorizationStatus: AuthorizationStatus.NO_AUTH,
  isDataLoaded: false,
  comments: {},
  preview: null,
};

const reducer = (state = initialState, action) => {
  switch (action.type) {
    case ActionType.SET_GENRE:
      return {
        ...state,
        genre: action.payload,
      };

    case ActionType.LOAD_SUCCESS:
      return {
        ...state,
        genres: createGenreArray(action.payload),
        films: action.payload,
        isDataLoaded: true,
      };

    case ActionType.LOAD_PROMO_FILM:
      return {
        ...state,
        preview: action.payload,
      };

    case ActionType.REQUIRED_AUTHORIZATION:
      return {
        ...state,
        authorizationStatus: action.payload,
      };

    case ActionType.SET_USER:
      return {
        ...state,
        user: action.payload,
      };

    case ActionType.ADD_REVIEW:
      const {comments} = state;

      return {
        ...state,
        comments: {...comments, [action.payload.id]: action.payload.comments},
      };

    case ActionType.LOAD_FAVORITE:
      return {
        ...state,
        favorite: action.payload,
      };

    case ActionType.SWITCH_FAVORITE_STATUS:
      if (state.preview && state.preview.id === action.payload.id) {
        state.preview = action.payload;
      }

      state.films = state.films.map((item) => item.id === action.payload.id ?
        action.payload : item
      );

      return {
        ...state,
        favorite: state.films.filter(({isFavorite}) => isFavorite),
      };
  }

  return state;
};


export {reducer};
