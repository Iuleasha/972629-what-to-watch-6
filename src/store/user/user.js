import {createReducer} from '@reduxjs/toolkit';
import {AuthorizationStatus} from '../../constants/constant';
import {requireAuthorization, setUser} from '../action';

const initialState = {
  authorizationStatus: AuthorizationStatus.NO_AUTH,
};

const user = createReducer(initialState, (builder) => {
  builder.addCase(requireAuthorization, (state, action) => {
    state.authorizationStatus = action.payload;
  });
  builder.addCase(setUser, (state, action) => {
    state.user = action.payload;
  });
});

export {user};
