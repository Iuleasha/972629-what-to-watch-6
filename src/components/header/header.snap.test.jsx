import {configureStore} from '@reduxjs/toolkit';
import {render} from '@testing-library/react';
import {createMemoryHistory} from 'history';
import React from 'react';
import {Provider} from 'react-redux';
import {Router} from 'react-router-dom';
import {adaptUserData} from '../../adapters/user';
import {AuthorizationStatus} from '../../constants/constant';
import {AUTH_INFO_MOCK} from '../../constants/mock';
import {requireAuthorization, setUser} from '../../store/action';
import reducer from '../../store/root-reducer';
import Header from './header';

const store = configureStore({reducer});

describe(`Test Header`, () => {
  const history = createMemoryHistory();

  test(`Should  render correctly with title`, () => {
    const {container} = render(<Provider store={store}><Router history={history}><Header
      title="Sign In"/></Router></Provider>);

    expect(container).toMatchSnapshot();
  });

  test(`Should render correctly with breadcrumbs`, () => {
    const {container} = render(<Provider store={store}><Router history={history}><Header
      breadcrumbs={{name: `Test`, id: `1`}}/></Router></Provider>);

    expect(container).toMatchSnapshot();
  });

  test(`Should render correctly with user`, () => {
    store.dispatch((dispatch) => {
      dispatch(setUser(adaptUserData(AUTH_INFO_MOCK)));
      dispatch(requireAuthorization(AuthorizationStatus.AUTH));
    });

    const {container} = render(<Provider store={store}><Router history={history}><Header/></Router></Provider>);

    expect(container).toMatchSnapshot();
  });
});
