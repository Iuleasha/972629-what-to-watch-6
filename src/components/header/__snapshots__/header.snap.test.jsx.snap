// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Test Header Should  render correctly with title 1`] = `
<div>
  <header
    class="page-header undefined"
  >
    <div
      class="logo"
    >
      <a
        class="logo__link"
        href="/"
      >
        <span
          class="logo__letter logo__letter--1"
        >
          W
        </span>
        <span
          class="logo__letter logo__letter--2"
        >
          T
        </span>
        <span
          class="logo__letter logo__letter--3"
        >
          W
        </span>
      </a>
    </div>
    <h1
      class="page-title user-page__title"
    >
      Sign In
    </h1>
    <div
      class="user-block"
    >
      <a
        class="user-block__link"
        href="/login"
      >
        Sign in
      </a>
    </div>
  </header>
</div>
`;

exports[`Test Header Should render correctly with breadcrumbs 1`] = `
<div>
  <header
    class="page-header undefined"
  >
    <div
      class="logo"
    >
      <a
        class="logo__link"
        href="/"
      >
        <span
          class="logo__letter logo__letter--1"
        >
          W
        </span>
        <span
          class="logo__letter logo__letter--2"
        >
          T
        </span>
        <span
          class="logo__letter logo__letter--3"
        >
          W
        </span>
      </a>
    </div>
    <nav
      class="breadcrumbs"
    >
      <ul
        class="breadcrumbs__list"
      >
        <li
          class="breadcrumbs__item"
        >
          <a
            class="breadcrumbs__link"
            href="/films/1"
          >
            Test
          </a>
        </li>
        <li
          class="breadcrumbs__item"
        >
          <a
            class="breadcrumbs__link"
          >
            Add review
          </a>
        </li>
      </ul>
    </nav>
    <div
      class="user-block"
    >
      <a
        class="user-block__link"
        href="/login"
      >
        Sign in
      </a>
    </div>
  </header>
</div>
`;

exports[`Test Header Should render correctly with user 1`] = `
<div>
  <header
    class="page-header undefined"
  >
    <div
      class="logo"
    >
      <a
        class="logo__link"
        href="/"
      >
        <span
          class="logo__letter logo__letter--1"
        >
          W
        </span>
        <span
          class="logo__letter logo__letter--2"
        >
          T
        </span>
        <span
          class="logo__letter logo__letter--3"
        >
          W
        </span>
      </a>
    </div>
    <div
      class="user-block"
    >
      <a
        class="user-block__avatar"
        href="/mylist"
      >
        <img
          alt="Oliver.conner"
          height="63"
          src="img/1.png"
          width="63"
        />
      </a>
      <a
        class="user-block__link"
        href="#"
      >
        Sign Out
      </a>
    </div>
  </header>
</div>
`;
